### PR TITLE
Fix double precision evaluation

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/ddp.py [hunk 1]
```diff
@@ -342,20 +342,20 @@ def training_step(self, *args, **kwargs) -> STEP_OUTPUT:
 
     def validation_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
         with self.precision_plugin.val_step_context():
-            if isinstance(self.model, DistributedDataParallel):
+            if self.lightning_module.trainer.state.fn == TrainerFn.FITTING:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/strategies/ddp.py:345]: Can you add this to the CHANGELOG too?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/strategies/ddp.py:345]: done!

File: CHANGELOG.md [hunk 2]
```diff
@@ -168,7 +168,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed an issue causing zero-division error for empty dataloaders ([#12885](https://github.com/PyTorchLightning/pytorch-lightning/pull/12885))
 
 
--
+- Fixed `Trainer(precision=64)` during evaluation ([#12983](https://github.com/PyTorchLightning/pytorch-lightning/pull/12983))
+
+
+- Fixed an issue to use wrapped `LightningModule` for evaluation during `trainer.fit` for `BaguaStrategy ([#12983](https://github.com/PyTorchLightning/pytorch-lightning/pull/12983))
```
Comment (carmocca) [inline CHANGELOG.md:174]:
```suggestion
- Fixed an issue to use wrapped `LightningModule` for evaluation during `trainer.fit` for `BaguaStrategy` ([#12983](https://github.com/PyTorchLightning/pytorch-lightning/pull/12983))
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/strategies/ddp_spawn.py [hunk 3]
```diff
@@ -241,20 +241,20 @@ def training_step(self, *args, **kwargs) -> STEP_OUTPUT:
...
     def validation_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
         with self.precision_plugin.val_step_context():
-            if isinstance(self.model, DistributedDataParallel):
+            if self.lightning_module.trainer.state.fn == TrainerFn.FITTING:
                 # used when calling `trainer.fit`
                 return self.model(*args, **kwargs)
             else:
                 # used when calling `trainer.validate`
-                return self.lightning_module.validation_step(*args, **kwargs)
+                return self.model.validation_step(*args, **kwargs)
```
Comment (awaelchli) [inline pytorch_lightning/strategies/ddp_spawn.py:249]: took me a moment to understand. 
here, model is not wrapped in DDP, but it's potentially wrapped in a double precision module, which is why we should reference self.model here. 

But this will only work if the wrapper implements the x_step methods, so this change here now introduces a typing issue. 

On a larger scope, now that the precision plugin is inside the strategy, I believe the precision plugin should be involved here directly instead of wrapping the model.

Comment (awaelchli) [reply pytorch_lightning/strategies/ddp_spawn.py:249]: #13000

Comment (rohitgr7) [reply pytorch_lightning/strategies/ddp_spawn.py:249]: yep. sounds reasonable.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/strategies/ddp.py (hunk 1 above)
Comment (four4fish) [inline pytorch_lightning/strategies/ddp.py:345]: How does this impact fsdp? fairscale fsdp is still inheritance from ddp right?

Comment (rohitgr7) [reply pytorch_lightning/strategies/ddp.py:345]: yes, but these methods are overridden there.

**Comment by rohitgr7** [comment]
gk

**Review by carmocca — APPROVED** [review]
